fuse/readonly/readonly_unix.go

```diff
@@ -133,31 +155,54 @@ func (s *Node) Attr(ctx context.Context, a *fuse.Attr) error {
 // Lookup performs a lookup under this node.
 func (s *Node) Lookup(ctx context.Context, name string) (fs.Node, error) {
 	log.Debugf("Lookup '%s'", name)
-	nodes, err := s.Ipfs.Resolver.ResolveLinks(ctx, s.Nd, []string{name})
-	if err != nil {
+	link, _, err := uio.ResolveUnixfsOnce(ctx, s.Ipfs.DAG, s.Nd, []string{name})
+	switch err {
+	case os.ErrNotExist, mdag.ErrLinkNotFound:
 		// todo: make this error more versatile.
 		return nil, fuse.ENOENT
+	default:
+		log.Errorf("fuse lookup %q: %s", name, err)
+		return nil, fuse.EIO
+	case nil:
+		// noop
 	}
 
-	pbnd, ok := nodes[len(nodes)-1].(*mdag.ProtoNode)
-	if !ok {
-		log.Error("fuse lookup got non-protobuf node")
-		return nil, fuse.ENOTSUP
+	nd, err := s.Ipfs.DAG.Get(ctx, link.Cid)
+	switch err {
+	case mdag.ErrNotFound:
+	default:
+		log.Errorf("fuse lookup %q: %s", name, err)
+		return nil, err
+	case nil:
+		// noop
 	}
 
-	return &Node{Ipfs: s.Ipfs, Nd: pbnd}, nil
+	return &Node{Ipfs: s.Ipfs, Nd: nd}, nil
 }
 
 // ReadDirAll reads the link structure as directory entries
 func (s *Node) ReadDirAll(ctx context.Context) ([]fuse.Dirent, error) {
 	log.Debug("Node ReadDir")
-	entries := make([]fuse.Dirent, len(s.Nd.Links()))
-	for i, link := range s.Nd.Links() {
-		n := link.Name
+	dir, err := uio.NewDirectoryFromNode(s.Ipfs.DAG, s.Nd)
+	if err != nil {
+		return nil, err
+	}
+
+	var entries []fuse.Dirent
+	err = dir.ForEachLink(ctx, func(lnk *format.Link) error {
+		n := lnk.Name
 		if len(n) == 0 {
-			n = link.Cid.String()
+			n = lnk.Cid.String()
 		}
-		entries[i] = fuse.Dirent{Name: n, Type: fuse.DT_File}
+		// TODO: calling everything a DT_File here might cause issues. But it
+		// will be expensive to query each child. However most shells call an
+		// additional 'stat' on each item in a directory listing, its probably
+		// okay.
+		entries = append(entries, fuse.Dirent{Name: n, Type: fuse.DT_File})
```

Comment:
As someone who has written programs that rely on this being accurate... :rage:. We may want to start inlining info like this into directories (but that's not very elegant).

Reply:
Also, if your argument is that most shells will call an additional stat, we might as well compute this up-front and cache it.

Reply:
Yeah, lets go ahead and do that.